test/state/host.cpp

```diff
@@ -0,0 +1,369 @@
+        acc.balance = 0;
+    }
+
+    // Register the destruction if not done already.
+    if (std::find(m_destructs.begin(), m_destructs.end(), addr) == m_destructs.end())
+    {
+        m_destructs.push_back(addr);
+        return true;
+    }
+    return false;
+}
+
+static address compute_new_address(const evmc_message& msg, uint64_t sender_nonce) noexcept
+{
+    hash256 addr_base_hash;
+    if (msg.kind == EVMC_CREATE)
+    {
+        const auto rlp_list = rlp::encode_tuple(address{msg.sender}, sender_nonce);
+        addr_base_hash = keccak256(rlp_list);
+    }
+    else
+    {
+        const auto init_code_hash = keccak256({msg.input_data, msg.input_size});
+        uint8_t buffer[1 + sizeof(msg.sender) + sizeof(msg.create2_salt) + sizeof(init_code_hash)];
+        static_assert(std::size(buffer) == 85);
+        buffer[0] = 0xff;
+        std::memcpy(&buffer[1], msg.sender.bytes, sizeof(msg.sender));
+        std::memcpy(
+            &buffer[1 + sizeof(msg.sender)], msg.create2_salt.bytes, sizeof(msg.create2_salt));
+        std::memcpy(&buffer[1 + sizeof(msg.sender) + sizeof(msg.create2_salt)],
+            init_code_hash.bytes, sizeof(init_code_hash));
+        addr_base_hash = keccak256({buffer, std::size(buffer)});
+    }
+    evmc_address new_addr{};
+    std::memcpy(new_addr.bytes, &addr_base_hash.bytes[12], sizeof(new_addr));
+    return new_addr;
+}
+
+std::optional<evmc_message> Host::prepare_message(evmc_message msg)
+{
+    auto& sender_acc = m_state.get(msg.sender);
+    const auto sender_nonce = sender_acc.nonce;
+
+    // Bump sender nonce.
+    if (msg.depth == 0 || msg.kind == EVMC_CREATE || msg.kind == EVMC_CREATE2)
+    {
+        if (sender_nonce == Account::NonceMax)
+            return {};  // Light early exception, cannot happen for depth == 0.
+        ++sender_acc.nonce;
+    }
+
+    if (msg.kind == EVMC_CREATE || msg.kind == EVMC_CREATE2)
+    {
+        // Compute and fill create address.
+        assert(msg.recipient == address{});
+        assert(msg.code_address == address{});
+        msg.recipient = compute_new_address(msg, sender_nonce);
+
+        // By EIP-2929, the  access to new created address is never reverted.
+        m_accessed_addresses.insert(msg.recipient);
+    }
+
+    return msg;
+}
+
+evmc::Result Host::create(const evmc_message& msg) noexcept
+{
+    assert(msg.kind == EVMC_CREATE || msg.kind == EVMC_CREATE2);
+
+    // Check collision as defined in pseudo-EIP https://github.com/ethereum/EIPs/issues/684.
+    // All combinations of conditions (nonce, code, storage) are tested.
+    if (const auto collision_acc = m_state.get_or_null(msg.recipient);
+        collision_acc != nullptr && (collision_acc->nonce != 0 || !collision_acc->code.empty()))
+        return evmc::Result{EVMC_OUT_OF_GAS};
+
+    auto& new_acc = m_state.get_or_create(msg.recipient);
+    if (m_rev >= EVMC_SPURIOUS_DRAGON)
+        new_acc.nonce = 1;
+    new_acc.storage.clear();  // In case of collision.
```

Comment:
confused why this is needed, doesn't it return error above in case of collision?

Reply:
The collision check ignores storage so you can have an "empty" account with non-empty storage (only in tests though). I have better version of this with better comments and I should bring it here.